src/compiler/checker.ts

```diff
@@ -15318,6 +15318,18 @@ namespace ts {
                         return getFlowTypeOfReference(node, getInferredClassType(classSymbol));
                     }
                 }
+                // Check if it's a constructor definition, can be either a variable decl or function decl
+                // i.e.
+                //   * /** @constructor */ function [name]() { ... }
+                //   * /** @constructor */ var x = function() { ... }
+                else if ((container.kind === SyntaxKind.FunctionExpression ||
+                            container.kind === SyntaxKind.FunctionDeclaration) &&
+                           getJSDocClassTag(container)) {
+                    const classType = getJavaScriptClassType(container.symbol);
+                    if (classType) {
+                        return getFlowTypeOfReference(node, classType);
+                    }
+                }

```

Comment:
I don't think this is the correct fix. The type of `this` should still be any because without an explicit or contextual type for this, you can still assign to any property on it. Instead, the no-implicit-this error should just be suppressed in javascript. The check at the bottom of `checkThisExpression` should change to `if (!type && noImplicitThis && !isInJavascriptFile(node)) {` so that JS-based uses of `this` in constructor functions won't error.